### PR TITLE
chore(scan-class): add new scan class

### PR DESCRIPTION
**Pull request by pedrooot**
### Description

New Scan class it's added. This class will allow the user to get the info related with a scan.

- [x] Tests

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: prowler/lib/check/check.py [hunk 1]
```diff
@@ -593,12 +593,17 @@ def execute_checks(
                     service,
                     check_name,
                     global_provider,
-                    services_executed,
-                    checks_executed,
                     custom_checks_metadata,
                 )
                 all_findings.extend(check_findings)
 
+                # Update Audit Status
```
Comment (jfagoagas) [inline prowler/lib/check/check.py:600]: Why are you moving this outside of the `execute` function? Is there any reason impacting the scan class?

File: prowler/lib/scan/scan.py [hunk 2]
```diff
@@ -1,60 +1,155 @@
...
-                    f"{check_name} - {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
-                )
-    except Exception as error:
-        logger.error(
-            f"{check_name} - {error.__class__.__name__}[{error.__traceback__.tb_lineno}]: {error}"
-        )
-
-    return all_findings
+class Scan:
+    # Maybe not needed
```
Comment (jfagoagas) [inline prowler/lib/scan/scan.py]:
```suggestion
```

File: prowler/lib/scan/scan.py [hunk 3]
```diff
@@ -1,60 +1,155 @@
...
+    """
+    get_service_name_from_check_name returns the service name for a given check name.
+
+    Example:
+        get_service_name_from_check_name("ec2_instance_public") -> "ec2"
+    """
+    return check_name.split("_")[0]
+
+
+def get_service_checks_to_execute(checks_to_execute: set[str]) -> dict[str, set[str]]:
```
Comment (jfagoagas) [inline prowler/lib/scan/scan.py:186]: Please add docstings for all the functions and methods.

File: tests/lib/scan/scan_test.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
+import unittest
+from unittest import mock
+
+from prowler.lib.scan.scan import Scan, get_service_checks_to_execute
+from tests.lib.outputs.fixtures.fixtures import generate_finding_output
+from tests.providers.aws.utils import set_mocked_aws_provider
+
+
+class Test_Scan(unittest.TestCase):
```
Comment (jfagoagas) [inline tests/lib/scan/scan_test.py]: Remember that we not use `unittest` we use `pytest`.

File: tests/lib/scan/scan_test.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+        # verify that update_audit_metadata was called twice
+        assert self.mock_update_audit_metadata.call_count == 2
+
+        # verify that the progress is 1.0
+        assert scan.progress == 1.0
+        assert scan._number_of_checks_completed == 2
+
+        # Verify that the findings are correct
+        assert scan.findings == self.findings + self.findings
```
Comment (jfagoagas) [inline tests/lib/scan/scan_test.py]: Why this?

File: tests/lib/scan/scan_test.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+        # Execute the scan
+        results = list(scan.scan(custom_checks_metadata))
+
+        # Verify that execute was called twice
+        assert self.mock_execute.call_count == 2
+
+        assert len(results) == 2
+        assert results[0][1] == self.findings
+        assert results[1][1] == self.findings
```
Comment (jfagoagas) [inline tests/lib/scan/scan_test.py]: Could you also assert the audit progress?

**Review by jfagoagas — CHANGES_REQUESTED** [review]
That's great @pedrooot 👏 

Please review my comments when you get a chance.

File: prowler/lib/check/check.py (hunk 1 above)
Comment (pedrooot) [reply prowler/lib/check/check.py:600]: Nop, it's just to modularize the code

File: prowler/lib/scan/scan.py (hunk 2 above)
Comment (pedrooot) [reply prowler/lib/scan/scan.py]: Done

File: prowler/lib/scan/scan.py (hunk 3 above)
Comment (pedrooot) [reply prowler/lib/scan/scan.py:186]: Done

File: tests/lib/scan/scan_test.py (hunk 4 above)
Comment (pedrooot) [reply tests/lib/scan/scan_test.py]: True, test have been refactored

File: tests/lib/scan/scan_test.py (hunk 5 above)
Comment (pedrooot) [reply tests/lib/scan/scan_test.py]: I've mocked the same finding twice, I could create a list that holds the two findings (wich are the same) or make it this way

File: tests/lib/scan/scan_test.py (hunk 6 above)
Comment (pedrooot) [reply tests/lib/scan/scan_test.py]: Sure! Done

**Review by jfagoagas — APPROVED** [review]
That's impressive @pedrooot 

Thanks for taking care of this great feature from start to finish 🏁 Great coding, documentation and tests 💯

**Pull request merged**